packages/graphql/src/services/practiceQuizzes.ts

```diff
@@ -27,32 +39,247 @@ export async function getPracticeQuizData(
 
   if (!quiz) return null
 
-  //   const elements = quiz.stacks.reduce((acc, stack) => {
-  //     return [...acc, ...stack.elements]
-  //   }, [])
-
-  //   console.log('practice quiz elements backend: ', elements)
-
-  // TODO: reintroduce ordering
-  // if (element.orderType === OrderType.LAST_RESPONSE) {
-  //   const orderedInstances = R.sort(
-  //     (a, b) => (a.lastResponse ?? 0) - (b.lastResponse ?? 0),
-  //     shuffle(instancesWithoutSolution.instances)
-  //   )
-
-  //   return {
-  //     ...element,
-  //     ...instancesWithoutSolution,
-  //     instances: orderedInstances,
-  //   }
-  // }
-
-  // if (element.orderType === OrderType.SHUFFLED) {
-  //   return {
-  //     ...element,
-  //     ...instancesWithoutSolution,
-  //     instances: shuffle(instancesWithoutSolution.instances),
-  //   }
-  // }
+  if (ctx.user?.sub && ctx.user?.role === UserRole.PARTICIPANT) {
+    // TODO: add time decay as well
+    // TODO: adapt the implementation to multiple instances per stack - resorting inside the stack does probably not make sense
+    const orderedStacks = quiz.stacks.sort((a, b) => {
+      const aResponses = a.elements[0].responses
+      const bResponses = b.elements[0].responses
+
+      // place instances, which have never been answered at the front
+      if (aResponses.length === 0 && bResponses.length === 0) return 0
+      if (aResponses.length === 0) return -1
+      if (bResponses.length === 0) return 1
+      const aResponse = aResponses[0]
+      const bResponse = bResponses[0]
+
+      // sort first according by correctCountStreak ascending, then by correctCount ascending, then by the lastResponseAt from old to new
+      if (aResponse.correctCountStreak < bResponse.correctCountStreak) return -1
+      if (aResponse.correctCountStreak > bResponse.correctCountStreak) return 1
+      if (aResponse.correctCount < bResponse.correctCount) return -1
+      if (aResponse.correctCount > bResponse.correctCount) return 1
+      if (aResponse.lastCorrectAt < bResponse.lastCorrectAt) return -1
+      if (aResponse.lastCorrectAt > bResponse.lastCorrectAt) return 1
+      return 0
+    })
+
+    return {
+      ...quiz,
+      stacks: orderedStacks,
+    }
+  }
   return quiz
 }
+
+interface RespondToFlashcardInstanceInput {
+  id: number
+  courseId: string
+  correctness: number
+}
+
+export async function respondToFlashcardInstance(
+  { id, courseId, correctness }: RespondToFlashcardInstanceInput,
+  ctx: Context
+) {
+  const existingInstance = await ctx.prisma.elementInstance.findUnique({
+    where: {
+      id,
+    },
+  })
+
+  if (!existingInstance) return null
+
+  // 0 = wrong, 1 = partial, 2 = correct
+  if (![0, 1, 2].includes(correctness)) {
+    return null
+  }
+
+  // TODO: also handle the case where the participant is logged in, but has no valid participation in the considered course
```

Comment:
maybe we can ignore participations and related topics completely, at least as long as the flashcards have no impact on leaderboards (just give XP)

Reply:
@rschlaefli In principle, I totally agree, but if I saw it correctly, the link from the question response table to the "Participation" table is required... Therefore, we either have to move away from this concept (which would probably require rigorous testing of all other use cases) or still require a participation in my opinion.